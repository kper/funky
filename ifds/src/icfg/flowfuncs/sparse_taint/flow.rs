use crate::icfg::state::State;
use crate::icfg::{flowfuncs::*, tabulation::sparse::Ctx};

pub struct SparseTaintNormalFlowFunction;

impl SparseTaintNormalFlowFunction {
    pub fn identity<'a>(
        &self,
        ctx: &mut Ctx<'a>,
        function: &AstFunction,
        fact: &Fact,
        pc: usize,
        defuse: &mut DefUseChain,
        edges: &mut Vec<Edge>,
    ) -> Result<()> {
        let after = defuse.get_next(ctx, function, &fact.belongs_to_var, pc)?;

        for to in after {
            edges.push(Edge::Normal {
                from: fact.apply(),
                to: to,
                curved: false,
            });
        }

        Ok(())
    }
}

impl SparseNormalFlowFunction for SparseTaintNormalFlowFunction {
    fn flow<'a>(
        &self,
        ctx: &mut Ctx<'a>,
        function: &AstFunction,
        pc: usize,
        variable: &String,
        defuse: &mut DefUseChain,
    ) -> Result<Vec<Fact>> {
        debug!(
            "Calling flow for {} with var {} with pc {}",
            function.name, variable, pc
        );

        let mut facts = Vec::new();

        let instructions = &function.instructions;
        let instruction = instructions.get(pc);

        if instruction.is_none() {
            debug!("Instruction is none");
            return Ok(Vec::new());
        }

        let instruction = instruction.unwrap();

        debug!("Next instruction is {:?} for {}", instruction, variable);

        let mut nodes = defuse.demand(ctx, function, variable, pc)?;

        match instruction {
            Instruction::Unop(dest, ..) |
            Instruction::Phi(dest, ..) |
            Instruction::BinOp(dest, ..) | 
            Instruction::Assign(dest, ..) => {
                let x = defuse.demand_inclusive(ctx, function, dest, pc)?;
                nodes.extend(x.into_iter().map(|x| x.clone()));
            }
            _ => {}
        }
        let nodes = nodes
            .into_iter()
            .filter_map(|x| {
                if x.pc != x.next_pc {
                    Some(x.apply())
                } else {
                    Some(x)
                }
            })
            .collect::<Vec<_>>();

        facts.extend(nodes.into_iter());

        debug!("Out {:#?}", facts);

        Ok(facts)
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::icfg::state::State;
    use crate::ir::ast::Program;
    use pretty_assertions::assert_eq;

    #[test]
    fn test_binop() {
        /*
            %0 = 1
            %1 = 1
            %2 = %0 op %1
            %3 = %1 op %0
        */

        let func_name = "main".to_string();
        let function = AstFunction {
            name: func_name.clone(),
            definitions: vec![
                "%0".to_string(),
                "%1".to_string(),
                "%2".to_string(),
                "%3".to_string(),
            ],
            instructions: vec![
                Instruction::Const("%0".to_string(), 1.0),
                Instruction::Const("%1".to_string(), 1.0),
                Instruction::BinOp("%2".to_string(), "%0".to_string(), "%1".to_string()),
                Instruction::BinOp("%3".to_string(), "%0".to_string(), "%1".to_string()),
            ],
            ..Default::default()
        };

        let mut graph = Graph::default();
        let mut state = State::default();

        let mut ctx = Ctx {
            graph: &mut graph,
            state: &mut state,
            prog: &Program {
                functions: vec![function.clone()],
            },
            block_resolver: HashMap::default(),
        };

        let pc = 0;

        // fullfilling precondition of `chain.cache()`
        ctx.state.init_function(&function, pc).unwrap();

        let mut defuse = DefUseChain::default();

        let sparse = SparseTaintNormalFlowFunction;

        let facts = sparse
            .flow(&mut ctx, &function, 2, &"%0".to_string(), &mut defuse)
            .unwrap();

        debug!("facts {:#?}", facts);
        assert_eq!(4, facts.len());

        let cmp_facts: Vec<Fact> = vec![
            Fact {
                belongs_to_var: "%0".to_string(),
                var_is_global: false,
                var_is_taut: false,
                var_is_memory: false,
                pc: 4,
                next_pc: 4,
                track: 1,
                function: "main".to_string(),
                memory_offset: None,
            },
            Fact {
                belongs_to_var: "%0".to_string(),
                var_is_global: false,
                var_is_taut: false,
                var_is_memory: false,
                pc: 4,
                next_pc: 4,
                track: 1,
                function: "main".to_string(),
                memory_offset: None,
            },
            Fact {
                belongs_to_var: "%2".to_string(),
                var_is_global: false,
                var_is_taut: false,
                var_is_memory: false,
                pc: 3,
                next_pc: 4,
                track: 3,
                function: "main".to_string(),
                memory_offset: None,
            },
            Fact {
                belongs_to_var: "%2".to_string(),
                var_is_global: false,
                var_is_taut: false,
                var_is_memory: false,
                pc: 4,
                next_pc: 4,
                track: 3,
                function: "main".to_string(),
                memory_offset: None,
            },
        ];
        assert_eq!(cmp_facts, facts);
    }
}
