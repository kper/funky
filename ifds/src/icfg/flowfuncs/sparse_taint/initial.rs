use crate::icfg::tabulation::sparse::defuse::DefUseChain;
use crate::icfg::{flowfuncs::*, tabulation::sparse::Ctx};

pub struct SparseTaintInitialFlowFunction;

impl SparseInitialFlowFunction for SparseTaintInitialFlowFunction {
    fn flow<'a>(
        &self,
        ctx: &mut Ctx<'a>,
        function: &AstFunction,
        pc: usize,
        init_facts: &Vec<Fact>,
        defuse: &mut DefUseChain,
    ) -> Result<Vec<Edge>> {
        debug!("Calling init flow for {} with pc {}", function.name, pc);

        let mut edges = Vec::new();

        // We need the offset, because not every
        // instruction is taintable. For example BLOCK and JUMP
        // have no registers. That's why skip to the next one.
        let offset = 0;
        let instructions = &function.instructions;

        let init_fact = init_facts.get(0).context("Cannot find taut")?.clone();

        debug!("init fact is {:#?}", init_fact);

        loop {
            let pc = pc + offset;
            let instruction = instructions.get(pc).context("Cannot find instr")?;
            debug!("Next instruction is {:?}", instruction);

            match instruction {
                Instruction::Const(dest, _)
                | Instruction::Unknown(dest)
                | Instruction::Assign(dest, _)
                | Instruction::Unop(dest, _)
                | Instruction::BinOp(dest, _, _)
                | Instruction::Phi(dest, _, _)
                | Instruction::Conditional(dest, _)
                | Instruction::Load(dest, _, _) => {
                    let before = vec![init_fact.clone()];

                    for b in before.into_iter() {
                        let after_taut = defuse
                            .demand(ctx, &function, &"taut".to_string(), b.pc)
                            .context("Cannot find taut's fact")?;

                        debug!("Next taut is {:#?}", after_taut);

                        for taut in after_taut.into_iter() {
                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: taut.clone(),
                            });
                        }

                        let after_var = defuse
                            .demand(ctx, &function, dest, pc)
                            .context("Cannot find var's fact")?
                            .into_iter()
                            .map(|x| x.clone())
                            .collect::<Vec<_>>();

                        // append all left sides to the nodes
                        // %2 = binop %0 %1 -- there %2 is the left side
                        let mut appended = Vec::new();
                        let mut append_lhs = |dest: &String| -> Result<()> {
                            defuse.force_remove_if_outdated(function, dest, pc)?;
                            let x = defuse.demand_inclusive(ctx, function, dest, pc)?;
                            appended.extend(x.into_iter().map(|x| x.clone()));

                            Ok(())
                        };

                        for var in after_var.clone().iter() {
                            append_lhs(&var.belongs_to_var)?;
                        }

                        for var in after_var.into_iter().chain(appended) {
                            let applied = var.apply();

                            // after
                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: applied.clone(),
                            });
                        }
                    }
                }
                Instruction::Call(_callee, _, dests) => {
                    // When the analysis starts at a Call Instruction,
                    // then the function doesn't get called.
                    // But all destinations get automatically tainted.

                    let before = vec![init_fact.clone()];
                    for b in before.into_iter() {
                        let after_taut = defuse
                            .demand(ctx, &function, &"taut".to_string(), b.pc)
                            .context("Cannot find taut's fact")?;

                        debug!("Next taut is {:#?}", after_taut);

                        for taut in after_taut.into_iter() {
                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: taut.clone(),
                            });
                        }
                    }
                    for dest in dests {
                        let after_var = defuse
                            .demand(ctx, &function, dest, pc)
                            .context("Cannot find var's fact")?
                            .into_iter()
                            .map(|x| x.clone())
                            .collect::<Vec<_>>();

                        // append all left sides to the nodes
                        // %2 = binop %0 %1 -- there %2 is the left side
                        let mut appended = Vec::new();
                        let mut append_lhs = |dest: &String| -> Result<()> {
                            defuse.force_remove_if_outdated(function, dest, pc)?;
                            let x = defuse.demand_inclusive(ctx, function, dest, pc)?;
                            appended.extend(x.into_iter().map(|x| x.clone()));

                            Ok(())
                        };

                        for var in after_var.clone().iter() {
                            append_lhs(&var.belongs_to_var)?;
                        }

                        for var in after_var.into_iter().chain(appended) {
                            let applied = var.apply();

                            // after
                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: applied.clone(),
                            });
                        }
                    }
                }
                Instruction::Store(_src, offset, _) => {
                    let before = vec![init_fact.clone()];

                    for b in before.into_iter() {
                        // The tautological fact was built by the `pacemaker`
                        // and will not be sparsely propagated.

                        let after_taut = defuse
                            .demand(ctx, &function, &"taut".to_string(), pc)
                            .context("Cannot find taut's fact")?;

                        for taut in after_taut.into_iter() {
                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: taut.clone(),
                            });
                        }

                        let mem = ctx
                            .state
                            .add_memory_var(function.name.clone(), offset.clone() as usize);

                        let after_var = defuse
                            .demand_inclusive(ctx, &function, &mem.name, pc)
                            .context("Cannot find var's fact")?;

                        for var in after_var.into_iter() {
                            let mut applied = var.apply();

                            assert!(applied.pc <= applied.next_pc);

                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: applied.clone(),
                            });
                        }
                    }
                }
                Instruction::Return(_dests) => {
                    let before = vec![init_fact.clone()];

                    for b in before.into_iter() {
                        // The tautological fact was built by the `pacemaker`
                        // and will not be sparsely propagated.

                        let after_taut = defuse
                            .demand(ctx, &function, &"taut".to_string(), pc)
                            .context("Cannot find taut's fact")?;

                        for taut in after_taut.into_iter() {
                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: taut.clone(),
                            });
                        }

                        let after_var = defuse
                            .demand(ctx, &function, &"taut".to_string(), pc)
                            .context("Cannot find var's fact")?;

                        for var in after_var.into_iter() {
                            let applied = var.apply();

                            assert!(applied.pc <= applied.next_pc);

                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: var.clone(),
                            });
                        }
                    }
                }
                Instruction::Block(_) | Instruction::Jump(_) => {
                    panic!("Block or Jump as first instruction is not supported");
                }
                _ => {}
            }

            break;
        }

        Ok(edges)
    }
}
