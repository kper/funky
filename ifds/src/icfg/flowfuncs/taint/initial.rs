use crate::icfg::flowfuncs::*;

pub struct TaintInitialFlowFunction;

impl InitialFlowFunction for TaintInitialFlowFunction {
    fn flow(
        &self,
        function: &AstFunction,
        graph: &mut Graph,
        pc: usize,
        init_facts: &Vec<Fact>,
        normal_flows_debug: &mut Vec<Edge>,
        state: &mut State,
    ) -> Result<Vec<Edge>> {
        debug!("Calling init flow for {} with pc {}", function.name, pc);

        let mut edges = Vec::new();

        // We need the offset, because not every
        // instruction is taintable. For example BLOCK and JUMP
        // have no registers. That's why skip to the next one.
        let mut offset = 0;
        let instructions = &function.instructions;

        let mut init_fact = init_facts.get(0).context("Cannot find taut")?.clone();

        loop {
            let pc = pc + offset;
            let instruction = instructions.get(pc).context("Cannot find instr")?;
            debug!("Next instruction is {:?}", instruction);

            state.add_statement(
                function,
                format!("{:?}", instruction),
                pc,
                &"taut".to_string(),
            )?;

            let _after2 = state
                .get_facts_at(&function.name, pc)?
                .filter(|x| x.belongs_to_var == "taut".to_string())
                .collect::<Vec<_>>();

            let after2 = _after2.get(0).unwrap();

            edges.push(Edge::Path {
                from: init_fact.clone().clone(),
                to: after2.clone().clone(),
            });

            match instruction {
                Instruction::Const(reg, _) => {
                    state.add_statement(function, format!("{:?}", instruction), pc + 1, reg)?;

                    let before2 = vec![init_fact.clone()];

                    let after2 = state
                        .get_facts_at(&function.name, pc + 1)?
                        .filter(|x| &x.belongs_to_var == reg)
                        .cloned();

                    for (b, a) in before2.into_iter().zip(after2) {
                        normal_flows_debug.push(Edge::Normal {
                            from: b.clone(),
                            to: a.clone(),
                            curved: false,
                        });
                        edges.push(Edge::Path {
                            from: init_fact.clone().clone(),
                            to: a.clone(),
                        });
                    }
                }
                Instruction::Assign(dest, _src) => {
                    let before2 = vec![init_fact.clone()];

                    state.add_statement(function, format!("{:?}", instruction), pc + 1, dest)?;

                    let after2 = state
                        .get_facts_at(&function.name, pc + 1)?
                        .filter(|x| &x.belongs_to_var == dest)
                        .cloned();

                    for (b, a) in before2.into_iter().zip(after2) {
                        normal_flows_debug.push(Edge::Normal {
                            from: b.clone(),
                            to: a.clone(),
                            curved: false,
                        });
                        edges.push(Edge::Path {
                            from: init_fact.clone().clone(),
                            to: a,
                        });
                    }
                }
                Instruction::Unop(dest, _src) => {
                    let before2 = vec![init_fact.clone()];

                    state.add_statement(function, format!("{:?}", instruction), pc + 1, dest)?;

                    let after2 = state
                        .get_facts_at(&function.name, pc + 1)?
                        .filter(|x| &x.belongs_to_var == dest)
                        .cloned();

                    for (b, a) in before2.into_iter().zip(after2) {
                        normal_flows_debug.push(Edge::Normal {
                            from: b.clone(),
                            to: a.clone(),
                            curved: false,
                        });
                        edges.push(Edge::Path {
                            from: init_fact.clone().clone(),
                            to: a,
                        });
                    }
                }
                Instruction::BinOp(dest, _src, _src2) => {
                    let before2 = vec![init_fact.clone()];

                    state.add_statement(function, format!("{:?}", instruction), pc + 1, dest)?;

                    let after2 = state
                        .get_facts_at(&function.name, pc + 1)?
                        .filter(|x| &x.belongs_to_var == dest)
                        .cloned();

                    for (b, a) in before2.into_iter().zip(after2) {
                        normal_flows_debug.push(Edge::Normal {
                            from: b.clone(),
                            to: a.clone(),
                            curved: false,
                        });
                        edges.push(Edge::Path {
                            from: init_fact.clone().clone(),
                            to: a,
                        });
                    }
                }
                Instruction::Conditional(reg, _) => {
                    let before2 = vec![init_fact.clone()];

                    state.add_statement(function, format!("{:?}", instruction), pc + 1, reg)?;

                    let after2 = state
                        .get_facts_at(&function.name, pc + 1)?
                        .filter(|x| x.belongs_to_var == "taut".to_string())
                        .cloned();

                    for (b, a) in before2.into_iter().zip(after2) {
                        normal_flows_debug.push(Edge::Normal {
                            from: b.clone(),
                            to: a.clone(),
                            curved: false,
                        });
                        edges.push(Edge::Path {
                            from: init_fact.clone().clone(),
                            to: a,
                        });
                    }
                }
                Instruction::Block(_) | Instruction::Jump(_) => {
                    let before2 = vec![init_fact.clone()];

                    state.add_statement(
                        function,
                        format!("{:?}", instruction),
                        pc + 1,
                        &"taut".to_string(),
                    )?;

                    let after2 = state
                        .get_facts_at(&function.name, pc + 1)?
                        .filter(|x| x.belongs_to_var == "taut".to_string())
                        .cloned();

                    // Replace the init_fact for the next iteration.
                    // Because, we would skip one row if not.
                    let after3 = state
                        .get_facts_at(&function.name, pc + 1)?
                        .filter(|x| x.belongs_to_var == "taut".to_string())
                        .cloned();

                    init_fact = after3.collect::<Vec<_>>().get(0).unwrap().clone();

                    for (b, a) in before2.into_iter().zip(after2) {
                        normal_flows_debug.push(Edge::Normal {
                            from: b.clone(),
                            to: a.clone(),
                            curved: false,
                        });
                        edges.push(Edge::Path {
                            from: init_fact.clone().clone(),
                            to: a,
                        });
                    }

                    offset += 1;

                    continue;
                }
                Instruction::Phi(dest, _src1, _src2) => {
                    let before2 = vec![init_fact.clone()];

                    state.add_statement(function, format!("{:?}", instruction), pc + 1, dest)?;

                    let after2 = state
                        .get_facts_at(&function.name, pc + 1)?
                        .filter(|x| &x.belongs_to_var == dest)
                        .cloned();

                    for (b, a) in before2.into_iter().zip(after2) {
                        normal_flows_debug.push(Edge::Normal {
                            from: b.clone(),
                            to: a.clone(),
                            curved: false,
                        });
                        edges.push(Edge::Path {
                            from: init_fact.clone().clone(),
                            to: a,
                        });
                    }
                }
                _ => {}
            }

            break;
        }

        Ok(edges)
    }
}
