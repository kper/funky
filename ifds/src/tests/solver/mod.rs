use crate::solver::*;

use crate::icfg::convert::ConvertSummary;
use crate::icfg::tikz::render_to;
use insta::assert_snapshot;

use crate::grammar::*;

use crate::icfg::flowfuncs::taint::flow::TaintNormalFlowFunction;
use crate::icfg::flowfuncs::taint::initial::TaintInitialFlowFunction;

macro_rules! ir {
    ($name:expr, $req:expr, $ir:expr) => {{
        let mut convert = ConvertSummary::new(TaintInitialFlowFunction, TaintNormalFlowFunction);

        let prog = ProgramParser::new().parse(&$ir).unwrap();

        let (graph, state) = convert.visit(&prog, &$req).unwrap();

        let output = render_to(&graph, &state);

        assert_snapshot!(format!("{}_dot", $name), output);

        graph
    }};
}

macro_rules! check_vars {
    ($name:expr, $solver:expr, $graph:expr, $req:expr) => {
        let mut vars = $solver
                    .sinks_var(&mut $graph, &$req)
                    .unwrap()
                    .into_iter()
                    .collect::<Vec<_>>();

        vars.sort();

        assert_snapshot!(
            format!("{}_vars", $name),
            format!(
                "{:#?}",
                vars 
            )
        );
    };
}

#[test]
fn test_intra_reachability() {
    let mut solver = IfdsSolver;

    let name = "intra_reach";

    let req = Request {
        function: "test".to_string(),
        pc: 0,
        variable: Some("%0".to_string()),
    };

    let mut graph = ir!(
        name,
        req,
        "
        define test (result 0) (define %0 %1) {
            %0 = 1
            %1 = %0
         };
    "
    );

    let req = Request {
        variable: Some("%0".to_string()),
        function: "test".to_string(),
        pc: 0,
    };
    let sinks = solver.all_sinks(&mut graph, &req).unwrap();

    assert_snapshot!(name, format!("{:#?}", sinks));

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 0,
                variable: Some("%0".to_string()),
            }
        )
        .unwrap());

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 1,
                variable: Some("%1".to_string()),
            }
        )
        .unwrap());

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 1,
                variable: Some("%0".to_string()),
            }
        )
        .unwrap());

    check_vars!(name, solver, graph, req);
}

#[test]
fn test_kill() {
    let mut solver = IfdsSolver;

    let name = "test_kill";

    let req = Request {
        function: "test".to_string(),
        pc: 0,
        variable: Some("%0".to_string()),
    };

    let mut graph = ir!(
        name,
        req,
        "
        define test (result 0) (define %0) {
            %0 = 1
            KILL %0
         };
    "
    );

    let req = Request {
        variable: Some("%0".to_string()),
        function: "test".to_string(),
        pc: 0,
    };
    let sinks = solver.all_sinks(&mut graph, &req).unwrap();

    assert_snapshot!(name, format!("{:#?}", sinks));

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 0,
                variable: Some("%0".to_string()),
            }
        )
        .unwrap());

    // should not be reachable anymore
    assert!(!solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 1,
                variable: Some("%1".to_string()),
            }
        )
        .unwrap());

    check_vars!(name, solver, graph, req);
}

#[test]
fn test_loop() {
    let mut solver = IfdsSolver;

    let name = "test_loop";

    let req = Request {
        function: "main".to_string(),
        pc: 1,
        variable: Some("%0".to_string()),
    };

    let mut graph = ir!(
        name,
        req,
        "
        define main (result 0) (define %0 %1) {
            BLOCK 0
            %0 = 1
            GOTO 0 
        };
    "
    );

    let sinks = solver.all_sinks(&mut graph, &req).unwrap();

    assert_snapshot!(name, format!("{:#?}", sinks));

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "main".to_string(),
                pc: 1,
                variable: Some("%0".to_string()),
            }
        )
        .unwrap());

    check_vars!(name, solver, graph, req);
}

#[test]
fn test_functions() {
    let mut solver = IfdsSolver;

    let name = "test_functions";

    let req = Request {
        function: "test".to_string(),
        pc: 0,
        variable: Some("%0".to_string()),
    };

    let mut graph = ir!(
        name,
        req,
        "
        define test (result 0) (define %0) {
            %0 = 1
            CALL mytest(%0)
        };
        define mytest (param %0) (result 0) (define %0 %1)  {
            %0 = 2   
            %1 = 3
            RETURN;
        };
    "
    );

    let sinks = solver.all_sinks(&mut graph, &req).unwrap();

    assert_snapshot!(name, format!("{:#?}", sinks));

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 0,
                variable: Some("%0".to_string()),
            }
        )
        .unwrap());

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 1,
                variable: Some("%0".to_string()),
            }
        )
        .unwrap());

    check_vars!(name, solver, graph, req);
}

#[test]
fn test_gcd() {
    let mut solver = IfdsSolver;

    let name = "test_gcd";

    let req = Request {
        function: "0".to_string(),
        pc: 17,
        variable: Some("%11".to_string()),
    };

    let mut graph = ir!(
        name,
        req,
        "
       define 0 (param %0 %1) (result 1) (define %0 %1 %2 %3 %4 %5 %6 %7 %8 %9 %10 %11 %12 %13 %14 %15 %16) {
        BLOCK 0
        BLOCK 1
        %4 = %0
        %5 = %1
        %6 = %5 op %4
        IF %6 THEN GOTO 3 ELSE GOTO 4
        BLOCK 3 
        %7 = %0
        %2 = %7
        KILL %7
        GOTO 2
        GOTO 4
        BLOCK 4
        BLOCK 5
        %8 = %1
        %9 = 0
        %10 = %0
        %11 = %1
        %12 = %11 op %10
        %13 = %12
        %2 = %12
        %14 = %13 op %12
        %3 = %14
        %15 = %1
        %16 = 0
        BLOCK 2
        RETURN %15;
        }; 
    "
    );

    let sinks = solver.all_sinks(&mut graph, &req).unwrap();

    assert_snapshot!(name, format!("{:#?}", sinks));

    check_vars!(name, solver, graph, req);
}

#[test]
fn test_globals() {
    let mut solver = IfdsSolver;

    let name = "test_globals";

    let req = Request {
        function: "test".to_string(),
        pc: 0,
        variable: Some("%0".to_string()),
    };

    let mut graph = ir!(
        name,
        req,
        "
        define test (result 0) (define %-1 %0 %2) {
            %0 = 1
            %-1 = %0 
            %2 <- CALL mytest()
        };
        define mytest (param) (result 1) (define %-1 %0 %1)  {
            %0 = 2   
            %1 = 3
            RETURN %-1;
        };
    "
    );

    let sinks = solver.all_sinks(&mut graph, &req).unwrap();

    assert_snapshot!(name, format!("{:#?}", sinks));

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 0,
                variable: Some("%0".to_string()),
            }
        )
        .unwrap());

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 1,
                variable: Some("%-1".to_string()),
            }
        )
        .unwrap());

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 2,
                variable: Some("%2".to_string()),
            }
        )
        .unwrap());

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 2,
                variable: Some("%0".to_string()),
            }
        )
        .unwrap());

    check_vars!(name, solver, graph, req);
}

#[test]
fn test_returned_value() {
    let mut solver = IfdsSolver;

    let name = "test_returned_values";

    let req = Request {
        function: "0".to_string(),
        pc: 8,
        variable: Some("%11".to_string()),
    };

    let mut graph = ir!(
        name,
        req,
        "
        define 0 (param %0) (result 1) (define %0 %1 %7 %8 %9 %10 %11 %12 %13 %14 %15 %16 %17) {
            BLOCK 0
            BLOCK 1
            BLOCK 3
            %1 = %0
            %7 = %0
            %8 = -1
            %9 = %8 op %7
            %10 <- CALL 0(%9)
            %11 = %0
            %12 = -2
            %13 = %12 op %11
            %14 <- CALL 0(%13)
            %15 = %14 op %13
            KILL %15
            KILL %14
            RETURN %0;
        }; 
    "
    );

    let sinks = solver.all_sinks(&mut graph, &req).unwrap();

    assert_snapshot!(name, format!("{:#?}", sinks));

    check_vars!(name, solver, graph, req);
}

#[test]
fn test_looped_param() {
    let mut solver = IfdsSolver;

    let name = "test_looped_param";

    let req = Request {
        function: "0".to_string(),
        pc: 1,
        variable: Some("%7".to_string()),
    };

    let mut graph = ir!(
        name,
        req,
        "
        define 0 (param %0) (result 1) (define %0 %1 %2 %3 %4 %5 %6 %7 %8 %9 %10 %11 %12 %13 %14 %15 %16 %17) {
            %1 = %0
            %7 = %0
            %9 = %8 op %7
            %10 <- CALL 0(%9)
            %11 = %0
            RETURN %0;
        };
    "
    );

    let sinks = solver.all_sinks(&mut graph, &req).unwrap();

    assert_snapshot!(name, format!("{:#?}", sinks));

    check_vars!(name, solver, graph, req);
}

#[test]
fn test_memory() {
    let mut solver = IfdsSolver;

    let name = "test_memory";

    let req = Request {
        function: "0".to_string(),
        pc: 0,
        variable: Some("%0".to_string()),
    };

    let mut graph = ir!(
        name,
        req,
        "
        define 0 (param %0) (result 0) (define %0 %1 %2 %3 %4 %5 %6 %7) {
            %0 = -12345
            STORE FROM %0 OFFSET 0 + %0 ALIGN 2 32
            %4 = 8
            %5 = LOAD OFFSET 0 + %4 ALIGN 0
            %6 = 8
            %7 = LOAD OFFSET 0 + %6 ALIGN 0
        };
    "
    );

    let sinks = solver.all_sinks(&mut graph, &req).unwrap();

    assert_snapshot!(name, format!("{:#?}", sinks));

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "0".to_string(),
                pc: 0,
                variable: Some("%0".to_string()),
            }
        )
        .unwrap());

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "0".to_string(),
                pc: 1,
                variable: Some("%0".to_string()),
            }
        )
        .unwrap());

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "0".to_string(),
                pc: 1,
                variable: Some("mem@0".to_string()),
            }
        )
        .unwrap());

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "0".to_string(),
                pc: 3,
                variable: Some("%5".to_string()),
            }
        )
        .unwrap());

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "0".to_string(),
                pc: 5,
                variable: Some("%7".to_string()),
            }
        )
        .unwrap());

    check_vars!(name, solver, graph, req);
}

#[test]
fn test_ir_multiple_functions() {
    let mut solver = IfdsSolver;

    let name = "test_multiple_functions";

    let req = Request {
        function: "test".to_string(),
        pc: 0,
        variable: Some("%0".to_string()),
    };

    let mut graph = ir!(
        name,
        req,
        "
        define test (result 0) (define %0 %1 %2) {
            %0 = 1
            %1 <- CALL mytest(%0)
            %2 <- CALL mytest(%0)
        };
        define mytest (param %0) (result 1) (define %0)  {
            RETURN %0;
        };
    "
    );

    let sinks = solver.all_sinks(&mut graph, &req).unwrap();

    assert_snapshot!(name, format!("{:#?}", sinks));

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 1,
                variable: Some("%1".to_string()),
            }
        )
        .unwrap());

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 2,
                variable: Some("%2".to_string()),
            }
        )
        .unwrap());

    check_vars!(name, solver, graph, req);
}
