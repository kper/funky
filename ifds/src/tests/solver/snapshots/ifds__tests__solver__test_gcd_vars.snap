---
source: src/tests/solver/mod.rs
expression: "format!(\"{:#?}\", vars)"

---
[
    "%11",
    "%12",
    "%13",
    "%14",
    "%2",
    "%3",
]
